### PR TITLE
Refactor - Extracts DataConsolidator class

### PR DESCRIPTION
**Pull request by MarcoCode**

### CODE REVIEW

**Comment by MarcoCode** [comment]
@BrunoChauvet @ouranos @manu-d

File: Gemfile [hunk 1]
```diff
@@ -3,7 +3,7 @@ source 'http://rubygems.org'
 # Add dependencies required to use your gem here.
 gem 'rails', '~> 4.2'
 
-gem 'maestrano-rails', '~> 1.0.0'
+gem 'maestrano-rails', '~> 1.0.3'
```
Comment (ouranos) [inline Gemfile]: Not related to this PR, but to the general structure

For gems/engine, you shouldn't use the `Gemfile` but the gemspec.
The `Gemfile` should only contain the `gemspec` line.

See:
https://github.com/plataformatec/devise/blob/master/Gemfile
https://github.com/maestrano/mno-enterprise/blob/master/core/Gemfile

Btw, looking at the gemspec, it should be split between dev and runtime dependencies!

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
+module Maestrano::Connector::Rails::Services
+  class DataConsolidator
+    def initialize(organization, entity_self, options)
+      @is_a_subentity = entity_self.class.included_modules.include?(Maestrano::Connector::Rails::Concerns::SubEntityBase)
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: I think `entity_self.is_a?(Maestrano::Connector::Rails::Concerns::SubEntityBase)` would work too

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      @opts = options
+    end
+
+    def consolidate_singleton(connec_entities, external_entities)
+      return {connec_entities: [], external_entities: []} if external_entities.empty? && connec_entities.empty?
+
+      idmap = @current_entity.class.find_or_create_idmap(organization_id: @organization.id)
+      # No to_connec, to_external and inactive consideration here as we don't expect those workflow for singleton
+
+      keep_external = if external_entities.empty?
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: Maybe add a comment to explain what keep_external is gonna do?
It's not obvious for a connector noob like me :)

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                        false
+                      elsif connec_entities.empty?
+                        true
+                      elsif @opts.key?(:connec_preemption)
+                        !@opts[:connec_preemption]
+                      else
+                        !connec_more_recent?(connec_entities.first, external_entities.first)
+                      end
+
+      if keep_external
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: external_id is used in both branches 
```ruby
external_id = @current_entity.class.id_from_external_entity_hash(external_entities.first)
```
that might make the code a bit more concise

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        idmap.update(name: @current_entity.class.object_name_from_connec_entity_hash(entity), connec_id: unfold_hash[:connec_id])
+        idmap.update(external_id: @current_entity.class.id_from_external_entity_hash(external_entities.first)) unless external_entities.empty?
+        return {connec_entities: [{entity: @current_entity.map_to_external(entity), idmap: idmap, id_refs_only_connec_entity: {}}], external_entities: []}
+      end
+    end
+
+    def consolidate_connec_entities(connec_entities, external_entities, references, external_entity_name)
+      connec_entities.map do |entity|
+        # Entity has been created before date filtering limit
+        next nil if before_date_filtering_limit?(entity, false) && !@opts[:full_sync]
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: `nil` is redundant
`next if ..` will do the same

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+        # Unfold the id arrays
+        # From that point on, the connec_entity contains only string of external ids
+        unfold_hash = Maestrano::Connector::Rails::ConnecHelper.unfold_references(entity, references, @organization)
+        entity = unfold_hash[:entity]
+        next nil unless entity # discard if at least one record reference is missing
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+          # Expecting find_or_create to be mostly a create
+          idmap = @current_entity.class.find_or_create_idmap(organization_id: @organization.id, name: @current_entity.class.object_name_from_connec_entity_hash(entity), external_entity: external_entity_name.downcase, connec_id: connec_id)
+          next map_connec_entity_with_idmap(entity, external_entity_name, idmap, id_refs_only_connec_entity)
+        end
+
+        # Expecting find_or_create to be mostly a find
+        idmap = @current_entity.class.find_or_create_idmap(external_id: entity['id'], organization_id: @organization.id, external_entity: external_entity_name.downcase, connec_id: connec_id)
+        idmap.update(name: @current_entity.class.object_name_from_connec_entity_hash(entity))
+
+        next nil if idmap.external_inactive || !idmap.to_external || (!@opts[:full_sync] && not_modified_since_last_push_to_external?(idmap, entity))
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+        # Check for conflict with entities from external
+        solve_conflict(entity, external_entities, external_entity_name, idmap, id_refs_only_connec_entity)
+      end.compact
+    end
+
+    def consolidate_external_entities(external_entities, connec_entity_name)
+      external_entities.map do |entity|
+        # Entity has been created before date filtering limit
+        next nil if before_date_filtering_limit?(entity) && !@opts[:full_sync]
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+        entity_id = @current_entity.class.id_from_external_entity_hash(entity)
+        idmap = @current_entity.class.find_or_create_idmap(external_id: entity_id, organization_id: @organization.id, connec_entity: connec_entity_name.downcase)
+
+        # Not pushing entity to Connec!
+        next nil unless idmap.to_connec
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+        # Not pushing to Connec! and flagging as inactive if inactive in external application
+        inactive = @current_entity.class.inactive_from_external_entity_hash?(entity)
+        idmap.update(external_inactive: inactive, name: @current_entity.class.object_name_from_external_entity_hash(entity))
+        next nil if inactive
+
+        # Entity has not been modified since its last push to connec!
+        next nil if !@opts[:full_sync] && not_modified_since_last_push_to_connec?(idmap, entity)
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        entity_id = @current_entity.class.id_from_external_entity_hash(entity)
+        idmap = @current_entity.class.find_or_create_idmap(external_id: entity_id, organization_id: @organization.id, connec_entity: connec_entity_name.downcase)
+
+        # Not pushing entity to Connec!
+        next nil unless idmap.to_connec
+
+        # Not pushing to Connec! and flagging as inactive if inactive in external application
+        inactive = @current_entity.class.inactive_from_external_entity_hash?(entity)
+        idmap.update(external_inactive: inactive, name: @current_entity.class.object_name_from_external_entity_hash(entity))
+        next nil if inactive
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: redundant `nil`

File: app/models/maestrano/connector/rails/services/data_consolidator.rb [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+
+    def map_external_entity_with_idmap(external_entity, connec_entity_name, idmap)
+      if @is_a_subentity
+        {entity: @current_entity.map_to(connec_entity_name, external_entity, idmap.last_push_to_connec.nil?), idmap: idmap}
+      else
+        {entity: @current_entity.map_to_connec(external_entity, idmap.last_push_to_connec.nil?), idmap: idmap}
+      end
+    end
+
+    def map_connec_entity_with_idmap(connec_entity, external_entity_name, idmap, id_refs_only_connec_entity)
```
Comment (ouranos) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: Is this more readable?
```ruby 
entity = if @is_a_subentity
  @current_entity.map_to(external_entity_name, connec_entity, idmap.last_push_to_external.nil?)
else
  @current_entity.map_to_external(connec_entity, idmap.last_push_to_external.nil?)
end
{entity: entity, idmap: idmap, id_refs_only_connec_entity: id_refs_only_connec_entity}
```

**Comment by ouranos** [comment]
@hedudelgado FYI, you might be interested by this type of refactoring

File: Gemfile (hunk 1 above)
Comment (MarcoCode) [inline Gemfile]: Thanks @ouranos, will change this in the next PR

File: app/models/maestrano/connector/rails/services/data_consolidator.rb (hunk 2 above)
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: Yep, changed

File: app/models/maestrano/connector/rails/services/data_consolidator.rb (hunk 3 above)
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: I'm a noob at everything :) Thanks for spotting this

File: app/models/maestrano/connector/rails/services/data_consolidator.rb (hunk 12 above)
Comment (MarcoCode) [inline app/models/maestrano/connector/rails/services/data_consolidator.rb]: Yes, thanks!

**Pull request merged**